### PR TITLE
[runtime] Fix stacktrace creation on llvmonly backend

### PR DESCRIPTION
**Pull request by alexanderkyte**
This got corlib exception tests passing. 

Without this test, the bitcode backend will make the stacktrace of an exception object into the stacktrace of the site where it is thrown. 

This does not follow the normal .NET semantics. In .NET, catching an exception causes the stacktrace to be truncated. Only the frames that lie between where the exception was thrown and where it was caught are supposed to end up in the stacktrace.

I do this here by making a system that determines the number of managed frames above the current method (https://github.com/mono/mono/pull/3166/files#diff-39d30c5616c9c238bfca3c7abd6df526R394). This is used to find the number of frames above the place where the exception is caught, and to remove that number of frames from above the caught frame in the stacktrace (https://github.com/mono/mono/pull/3166/files#diff-39d30c5616c9c238bfca3c7abd6df526R3899). 

The other change lumped in here that is implicit in count_frames and build_stack_trace is that we now filter the wrapper frames out of the stack when it's created. This is to be expected, managed stacktraces shouldn't show a lot of m2n and n2m calls between the managed frames. 

The tests that this was written to fix (corlib's ExceptionTest) should be able to spot regressions or failures if this breaks in the future.

### CODE REVIEW

File: mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -8247,7 +8247,7 @@ emit_llvm_file (MonoAotCompile *acfg)
 	if (acfg->aot_opts.llvm_only) {
 		/* Use the stock clang from xcode */
 		// FIXME: arch
-		command = g_strdup_printf ("clang -march=x86-64 -fpic -msse -msse2 -msse3 -msse4 -O2 -fno-optimize-sibling-calls -Wno-override-module -c -o \"%s\" \"%s.opt.bc\"", acfg->llvm_ofile, acfg->tmpbasename);
+		command = g_strdup_printf ("clang -march=x86-64 -fpic -msse -msse2 -msse3 -msse4 -O0 -fno-optimize-sibling-calls -Wno-override-module -c -o \"%s\" \"%s.opt.bc\"", acfg->llvm_ofile, acfg->tmpbasename);
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: This is an unrelated change.

Comment (alexanderkyte) [inline mono/mini/aot-compiler.c]: Oops. Dirtied my WIP branch.

File: mono/mini/mini.h [hunk 2]
```diff
@@ -1118,6 +1118,20 @@ typedef struct {
...
+	/* 
+	 * The current exception in flight
+	 */
+	guint32 thrown_exc;
+
+	/* 
+	 * A temporary reference to the ips for the whole
+	 * stack trace when thrown. These are truncated when the
+	 * method is finally caught. 
+	 */
```
Comment (vargaz) [inline mono/mini/mini.h]: a typo.

File: mono/mini/mini-exceptions.c [hunk 3]
```diff
@@ -3068,9 +3077,8 @@ void
 mono_llvm_clear_exception (void)
 {
 	MonoJitTlsData *jit_tls = mono_get_jit_tls ();
-	mono_gchandle_free (jit_tls->thrown_exc);
-	jit_tls->thrown_exc = 0;
-
+	mono_gchandle_free (jit_tls->llvm_exc.thrown_exc);
+	memset (&jit_tls->llvm_exc, NULL, sizeof (MonoLLVMExceptionData));
 	mono_memory_barrier ();
```
Comment (vargaz) [inline mono/mini/mini-exceptions.c]: It should be a 0, not NULL.

File: mono/mini/mini-exceptions.c [hunk 4]
```diff
@@ -3123,9 +3131,58 @@ mono_llvm_match_exception (MonoJitInfo *jinfo, guint32 region_start, guint32 reg
...
+			size++;
+		}
+		/* The list contains ip-gshared info pairs */
+		size = size * 2;
+
+		MonoError error;
+		mono_error_init (&error);
+		MonoArray *trace_arr = mono_array_new_checked (mono_domain_get (), mono_defaults.int_class, size, &error);
+		return_val_if_nok (&error, NULL);
+		mono_error_cleanup(&error);
```
Comment (vargaz) [inline mono/mini/mini-exceptions.c]: No need to cleanup the error if its ok.

Comment (alexanderkyte) [inline mono/mini/mini-exceptions.c]: Actually, I think it probably makes more sense to assert here.

**Comment by alexanderkyte** [comment]
Hmm, the failures all look related to networking / sockets. I don't think that stacktrace string manipulation has anything to do with that.

**Comment by akoeplinger** [comment]
@alexanderkyte yes, those are flaky tests that we're tracking down, ignore them.

**Comment by alexanderkyte** [comment]
exceptions-18.exe is failing due to stacktrace issues on bitcode with this cherry-picked. Hold off on merging until I can diagnose.

**Comment by luhenry** [comment]
@vargaz could you please re-review?

@alexanderkyte could you please document in the PR description what this fixes exactly? And how do we verify / test that we don't regress?

**Comment by alexanderkyte** [comment]
@luhenry so this problem has test coverage through CI. 

That said, I'm not sure how our regression tests will catch bitcode breakages when the lane is abandoned https://jenkins.mono-project.com/job/test-mono-mainline-linux-bitcode/
